pkg/controllers/management/kontainerdrivermetadata/data.go

```diff
@@ -387,97 +401,260 @@ func getWindowsName(str string) string {
+			}
+			setting, err = md.Settings.Get(key, metav1.GetOptions{})
+			if err != nil {
+				return nil, false, fmt.Errorf("driverMetadata: error getting setting %s: %v", key, err)
+			}
+		}
+		if val, ok := setting.Labels[setting2.UserUpdateLabel]; ok && convert.ToString(val) == "true" {
+			userSettings[key] = get(key)
+		}
+	}
+	logrus.Debugf("driverMetadata: userSettings %v", userSettings)
+	if len(userSettings) > 0 {
+		return userSettings, true, nil
+	}
+	return userSettings, false, nil
+}
+
+func toUpdate(maxVersionForMajorK8sVersion map[string]string, deprecated map[string]bool,
+	defaultK8sVersions map[string]string, rancherVersion string, k8sVersionServiceOptions map[string]v3.KubernetesServicesOptions) (map[string]string, error) {
+
+	var k8sVersionsCurrent []string
+	var maxVersions []string
+	for k, v := range maxVersionForMajorK8sVersion {
+		if !deprecated[k] {
+			k8sVersionsCurrent = append(k8sVersionsCurrent, v)
+			maxVersions = append(maxVersions, k)
+		}
+	}
+	if len(maxVersions) == 0 {
+		return nil, fmt.Errorf("driverMetadata: no max version %v", maxVersionForMajorK8sVersion)
+	}
+	sort.Strings(k8sVersionsCurrent)
+	sort.Strings(maxVersions)
+
+	defaultK8sVersion, err := getDefaultK8sVersion(defaultK8sVersions, k8sVersionsCurrent, rancherVersion)
+	if err != nil {
+		return nil, err
+	}
+
 	k8sVersionRKESystemImages := map[string]interface{}{}
 	k8sVersionSvcOptions := map[string]v3.KubernetesServicesOptions{}
 
 	for majorVersion, k8sVersion := range maxVersionForMajorK8sVersion {
 		if !deprecated[k8sVersion] {
 			k8sVersionRKESystemImages[k8sVersion] = nil
-			k8sVersionSvcOptions[k8sVersion] = K8sVersionServiceOptions[majorVersion]
+			k8sVersionSvcOptions[k8sVersion] = k8sVersionServiceOptions[majorVersion]
 		}
 	}
 
-	var keys []string
-	for k := range maxVersionForMajorK8sVersion {
-		if !deprecated[k] {
-			keys = append(keys, k)
-		}
+	k8sCurrRKEdata, err := marshal(k8sVersionRKESystemImages)
+	if err != nil {
+		return nil, err
 	}
-	sort.Strings(keys)
 
-	return SaveSettings(k8sVersionRKESystemImages, k8sVersionSvcOptions, DefaultK8sVersions, rancherVersion, keys, deprecated)
-}
-
-func SaveSettings(k8sCurrVersions map[string]interface{},
-	k8sVersionSvcOptions map[string]v3.KubernetesServicesOptions, rancherDefaultK8sVersions map[string]string,
-	rancherVersion string, maxVersions []string, deprecated map[string]bool) error {
+	k8sSvcOptionData, err := marshal(k8sVersionSvcOptions)
+	if err != nil {
+		return nil, err
+	}
 
-	k8sCurrVersionData, err := marshal(k8sCurrVersions)
+	deprecatedData, err := marshal(deprecated)
 	if err != nil {
-		return err
+		return nil, err
+	}
+
+	minVersion := maxVersions[0]
+	maxVersion := util.GetTagMajorVersion(defaultK8sVersion)
+	uiSupported := fmt.Sprintf(">=%s.x <=%s.x", minVersion, maxVersion)
+	uiDefaultRange := fmt.Sprintf("<=%s.x", maxVersion)
+
+	return map[string]string{
+		"k8s-versions-current":            strings.Join(k8sVersionsCurrent, ","),
```

Comment:
shall we put all the keys in the constants somewhere in settings? it's being referenced in settings.go already I believe